### PR TITLE
Fix EBUSY errors under overlayfs and v4.13+ kernels

### PR DESCRIPTION
**Pull request by euank**
This removes and recreates the merged dir with each umount/mount
respectively.
This is done to make the impact of leaking mountpoints have less
user-visible impact.

It's fairly easy to accidentally leak mountpoints (even if moby doesn't,
other tools on linux like 'unshare' are quite able to incidentally do
so).

As of recently, overlayfs reacts to these mounts being leaked (see

One trick to force an unmount is to remove the mounted directory and
recreate it. Devicemapper now does this, overlay can follow suit.

**- Description for the changelog**

Fix upperdir in use warnings under overlayfs and v4.13+ kernels

**- A picture of a cute animal (not mandatory but encouraged)**

red panda </img>

### CODE REVIEW

**Comment by euank** [comment]
cc @rhvgoyal as the author of https://github.com/moby/moby/pull/22069, which this reverts, and @tonistiigi as the author of https://github.com/moby/moby/pull/27609, which I think makes the #22069 unnecessary.

**Comment by euank** [comment]
The powerpc failure is a flake. I don't know how to kick the jenkins to re-run

**Comment by thaJeztah** [comment]
@euank Looks like you forgot to sign-off the second commit 😅 

also ping @dmcgowan @rhvgoyal @tonistiigi PTAL

**Comment by rhvgoyal** [comment]
cc @amir73il 

Ok, so kernel commit `2cac0c00a6cdcc9` started giving errors in case of mount point leaks. Trying to avoid mount point leaks completely will make sense only if mount propagation was synchronous. That is by the time unmount finished, all the propagations and unmounting must have finished before it. If not, then we still have races where a leaked mount has not been unmounted yet (because propagation has not happened yet) and we are trying to create a mount point using same upper directory. I don't know if unmount is synchronous or not.

Secondly, even if you make `slave`, we will still have the race because runc makes all mount points rprivate by default and leaves a small race window where your unmount will not propagate.

Making it `slave/shared` in a way makes it more likely that we will leak mount points. Think of an application just doing `unshare -m` and now all future mounts under /var/lib/docker/overlay2/ will propagate to this application. 

I feel we can try to minimize mount point leaks but being able to take care of all corner cases will be very hard. So I am wondering if a solution should come from kernel side instead.

I am wondering will it make sense to convert kernel error message to a warning message? Or is there a way where kernel can figure out that we are essentially using same set of lower/upper/work directories and instead of instantiating a new super block, it re-uses existing super block (which is around due to busy mount). Amir will know much more, I am ccing him.

**Comment by amir73il** [comment]
@rhvgoyal overlas is right to complain, so kernel warning is not good enough.
@tonistiigi The root cause of the problem IIUC is the chroot() does not Unmount
all the tree under pivotDir.
You see making all mounts private in the new ns is just the first step.
The cloned overlay mounts, even though private are still in use by the new ns,
so they must be unmounted after pivot_root.
The Mount syscall below is not needed because all mounts in new ns are already private.
The Umount syscall below is wrong because it uses MNT_DETACH instead of
unmounting all tree below pivotDir synchronically.

 	// Make the pivotDir (where the old root lives) private so it can be unmounted without propagating to the host
 	if err := syscall.Mount("", pivotDir, "", syscall.MS_PRIVATE|syscall.MS_REC, ""); err != nil {
 		return fmt.Errorf("Error making old root private after pivot: %v", err)
 	}
 
 	// Now unmount the old root so it's no longer visible from the new root
 	if err := syscall.Unmount(pivotDir, syscall.MNT_DETACH); err != nil {
 		return fmt.Errorf("Error while unmounting old root after pivot: %v", err)
 	}

**Comment by euank** [comment]
@rhvgoyal

> Trying to avoid mount point leaks completely will make sense only if mount propagation was synchronous. 

That's not true. I can reliably reproduce a mount point leak to runc (leaked for 100ms to 1s) without this patch. With this patch I can't. It therefore still makes sense to try and avoid it.
Just because we can't be perfect isn't a reason to do nothing.

> I don't know if unmount is synchronous or not.

Because the code here uses MNT_DETACH as of https://github.com/moby/moby/pull/33638 we're opting in to async behavior either way. I think that PR is technically wrong, but that's something to address separate from this.
I do agree this code is still wrong in theory. I'm simply fixing the instance that is hit reliably in practice.

> Secondly, even if you make slave, we will still have the race because runc makes all mount points rprivate by default.

False. runc makes the mountpoints rslave. See either my original issue I say this fixes and my discussion there or my pr against [runc](https://github.com/opencontainers/runc/pull/1598#issue-259906691) and the referenced defaulting code.

> Making it slave/shared in a way makes it more likely that we will leak mount points. Think of an application just doing unshare -m and now all future mounts under /var/lib/docker/overlay2/ will propagate to this application.

False. `unshare -m` by default also sets `private` on `/`, so what you say will happen won't. An `unshare -m` someone runs won't get future mounts either before or after this change. It will break things either way, but I'm trying to fix runc stomping on docker, not arbitrary external users stomping on docker. Before or after this change, external users that don't use `rslave` will break docker roughly equally badly I think.

If they do `unshare -m --propagation slave` then they will get new mounts, but the `umounts` from the host will also propagate, so it's not a big deal. That mountpoint isn't leaked unless the program inside of `unshare -m` then remounts it private.

> I feel we can try to minimize mount point leaks but being able to take care of all corner cases will be very hard. So I am wondering if a solution should come from kernel side instead.

A better solution would be something like [@cyphar suggests](https://github.com/moby/moby/issues/34672#issuecomment-331569549) imo.
I think this current change is an incremental improvement that is necessary to fix real issues though.

> Or is there a way where kernel can figure out that we are essentially using same set of lower/upper/work directories and instead of instantiating a new super block, it re-uses existing super block (which is around due to busy mount).

A good first step there would be to not do "mount umount mount umount mount" to start a container, but instead just mount it once and leave it mounted.
That's also a significantly more painful refactor to do when there's already bleeding.

---------------------------------------------

@amir73il

> The root cause of the problem IMO is the chroot() does not Unmount
all the tree under pivotDir.

Read my original issue (https://github.com/moby/moby/issues/34672) where I talk about the actual issue I'm trying to solve. There may be more issues, but the issue I'm trying to fix does not actually use any of that code so it's not closely related to what I see as the root problem.

I agree the things you pointed out are problems (though I think the solution is making it `rslave` not `rprivate`), but I don't think that's related to getting this merged or the specific issue I'm trying to fix.

**Comment by amir73il** [comment]
Yeh I am confused of all the different issues and fixes. Anyway, the solution of rslave and lazy umount in cloned ns looks good.

**Comment by euank** [comment]
@amir73il 
> Yeh I am confused of all the different issues and fixes. Anyway, the solution of rslave and lazy umount in cloned ns looks good.

This PR is for fixing something else though. This is to fix the case where mounting `/var/lib/overlay2` as private explicitly means that an `unshare -m --propagation rslave`, such as the one done by runc, will have a *private* reference to `/var/lib/overlay2/$id/merged`, meaning that the overlay2 driver leaks the mount until the program with a cloned mountns exits.

Are you willing to review *this* change, or do you have specific questions about *this* issue that isn't answered already by my comments in https://github.com/moby/moby/issues/34672?

**Comment by amir73il** [comment]
I can only ACK the change that doesn't make home dir private. I don't know enough on the bug picture to say if var/lib/docker/overlay is always shared to begin with.

I suppose if makes little sense to make home dir private for other graph drivers? And the rslave change should be made to chroot() as well. Otherwise this fix is correct but partial. You shouldn't fix just the problem you see in front of you when you know there is a bigger issue to solve. But fine with me of this is the first step.

**Comment by euank** [comment]
@amir73il I apologize if I came off a little roughly. The internet does make it hard to convey tone properly.

> I can only ACK the change that doesn't make home dir private

Thanks! What I'm looking for here is making sure this change is in the right direction

> ... var/lib/docker/overlay is always shared to begin with.

I think it generally will be, especially in a systemd world. It's possible this should be swapping `MakePrivate` over to `MakeShared`, though I'm a little more wary about explicit shared than defaulting shared

> I suppose if makes little sense to make home dir private for other graph drivers?

It might, but I'm not an expert. I do know there are some related issues floating around for devicemapper at the least.

> And the rslave change should be made to chroot() as well

Yup, happy to look at that and changing the MNT_DETACH bit later. One step at a time.

**Comment by euank** [comment]
Review bump for this change

**Comment by rhvgoyal** [comment]
@amir73il Thinking more about it, it feels like regression to me (from userspace point of view). Sure, we can try modifying user space to not leak mount points, but kernel upgrade will still break existing container runtime.

And making sure setup is completely right and none of the mounts point are leaking is hard. 

So if I take a device `/dev/foo` and mount it on `/mnt/test`, leak the mount point, unmount it and mount again, it still works. It most likely gets same superblock.

IMHO, either we need to implement same behavior for overlay or reduce the level from error to warning.

And this is irrespective of docker changes. Sure try to reduce the amount of leaked mount points, that can only help. But it still feels like a regression.

**Comment by rhvgoyal** [comment]
@amir73il In some ways it fails like DOS to me. If an overlay mount point has leaked somewhere, then root can not create a new overlay mount point and get to its data.

Terminal 1
---------------
`mount -t overlay -o workdir=work,lowerdir=lower,upperdir=upper none merged/`

Terminal 2
--------------
`unshare -m`

Terminal 1
--------------
`umount merged`
`mount -t overlay -o workdir=work,lowerdir=lower,upperdir=upper none merged/`
`mount: /root/overlay-testing/merged: none already mounted or mount point busy`

Now root can't access its own data due to leaked mount point in some process mount namespace. And that sounds not desirable to me (given how easy it is to leak mount points).

**Comment by amir73il** [comment]
@rhvgoyal wrt regression you are definitely right. This should be fixed in stable kernels.
I will post a patch to make this new enforcement depend on new config option INDEX=y.

**Comment by tonistiigi** [comment]
cc @kolyshkin

**Comment by kolyshkin** [comment]
In my limited testing, this patch indeed helps with EBUSY on /merged on removal, which I can reproduce easily on a RHEL 7.4 system with `sysfs fs.may_detach_mounts = 0` and 100-200 short-lived containers starting/ending in parallel (`docker run --rm busybox true`).

This is not ultimate fix though as I still see occasional EBUSY on /shm removal (which is somewhat 
 harder to reproduce).

**Comment by euank** [comment]
@kolyshkin For the reasons described in #34672, I also had to apply this runc patch to fix my issue: https://github.com/opencontainers/runc/pull/1598

Is there a separate issue or bugzilla filed for the `/shm` ebusy you get with further details? It's possible that runc patch could help further, or that other people could help investigate.

Also, review ping! It seems like this PR has stalled out, but I don't think I've gotten any actionable requests / reasons why it's stalled.

@dmcgowan and @rhvgoyal, want to review this change?

Even if overlayfs lets us get away with leaking mounts, I don't think we should be and so IMO this should still be merged.

**Comment by doits** [comment]
I can confirm that this patch (applied to docker 17.10) fixes issue #34672. +1 for merging it.

**Comment by cpuguy83** [comment]
Why wouldn't we use rslave here?
Also wondering if https://github.com/moby/moby/pull/35217 is relevant (or correct) as well.

**Comment by euank** [comment]
@cpuguy83 rslave doesn't fix the problem I'm trying to fix.
I've created a simplified model of the problem [here](https://gist.github.com/euank/3cafdd8a318b1722f3d84f1d962272ac#file-repro-example-sh).

In that model, I've currently got it using `--make-rslave` for the simulated `/var/lib/docker/overlay2` directory.

On a 4.13 kernel, that will either fail, or on newer 4.13 it'll give a scary warning in dmesg and succeed.

However, on any kernel you can see that the mounts shown after "c2: still in runc init, current view of mounts" include the same c1 mount with the same mount ID, not a new mount with a different mount ID.

That's the leak I'm trying to squash.

Using rslave doesn't fix that leak because the leak is really two layers deep (runc has its own rslave subtree), and thus the umount sitll doesn't forward since slave mounts are only defined to receive from their master, not propagate as shared do.

If we switch it to explicitly be `MakeShared` vs Private or Rslave, that will work just fine. However, that's really what just skipping the `--make-private` does in practice. If you think it's cleaner to explicitly use `MakeShared` here, I could switch it to that rather than reverting the `MakePrivate` commit.

**Comment by cpuguy83** [comment]
Holy hell, you are right....
With using `rslave`, any new mounts you create underneath that dir in the fs tree will be `private`.


```
# mkdir -p /foo/{lower,upper,work,merged}
# mount --bind /foo /foo
# mount --make-rslave /foo
# mount -t overlay overlay -o upperdir=/foo/upper,lowerdir=/foo/lower,workdir=/foo/work /foo/merged
# findmnt -o TARGET,PROPAGATION /foo/merged
TARGET      PROPAGATION
/foo/merged private
```

With this, if I setup a new mount ns and then try to unmount `/foo/merged` from the parent, it does not propagate to the child.

It looks like if we leave it shared and then set the the actual overlay mount to slave the unmount propagates.

Testing this on `ubuntu-xenial 4.4.0-96-generic`, btw.

File: daemon/graphdriver/overlay2/overlay.go
Comment (euank) [inline daemon/graphdriver/overlay2/overlay.go]: Note, in the case of a live-restore upgrade, the previous umount won't be hit.. and after the upgrade to this new codepath, we now won't be cleaning up a private bindmount the previous version of dockerd setup.

Restarting the machine will eventually clean up this bindmount. I'm not sure if this cleanup code should keep unmounting and just ignore errors to handle post-live-restore cases, or whether it's okay to ignore the trouble since it's an edge case and doesn't cause too many ill effects

**Comment by cyphar** [comment]
@euank The core issue here is that mount propagation just simply isn't safe with namespaces, something that Eric Biederman and other kernel maintainers have known for a while -- the features were developed in parallel a long time ago. I've been [having discussions with them on the best way of resolving these problems from the kernel side](https://marc.info/?l=linux-xfs&m=150805760727451&w=2) (there are a huge variety of ways to cause a DoS of devicemapper and any other various filesystem mounts).

But back to this bug. Have you considered that something more akin to #34573 could solve this problem? Mount propagation is very fickle, because any unprivileged user can start changing it just by doing an `unshare -rm`. If we go one step further, and also remove the `lowerdir` or `upperdir` of a container, then `overlayfs` will do a cleanup of all mounts (even bind-mounts I believe -- which is something that `devicemapper` can't do).

**Comment by euank** [comment]
@cyphar 

If we could remove the upperdir in `Cleanup`, I think that would solve this certainly.
Unfortunately, we can't remove the upperdir without losing the container's data; docker is intentionally re-using the same mounts because it's the same container with the same data.

Am I misunderstanding the suggestion?

**Comment by rhvgoyal** [comment]
Now @amir73il has changed kernel to warn instead of error out (with index=off). So with that fix, atleast docker should get going again.

Still sorting out how to avoid mount point leaks is probably a good thing.

**Comment by cyphar** [comment]
@euank Right, removing the `upperdir` will lose data. In which case we should then just `rmdir` the mountpoint after `umount` -- which is what #34573 does for devicemapper. This will kill the mount if it's leaked to other namespaces. There are situation which this won't help with (such as a malicious user which intentionally creates a private bind-mount) but as I said, this is a kernel-side issue that is being worked on.

**Comment by kolyshkin** [comment]
> Right, removing the upperdir will lose data. In which case we should then just rmdir the mountpoint after > umount -- which is what #34573 does for devicemapper

@cyphar @euank @dmcgowan do we want to implement a similar thing for overlay2?

**Comment by euank** [comment]
@kolyshkin, @cyphar  Mind clarifying your question/suggestion?

What i was trying to say is that I don't think we can do what devicemapper is doing. It's okay in devicemapper to delete the mountpoint and create a new one since the data lives elsewhere.
Additional mounts can be created with the same underlying data as a result.

In overlay, the data is coupled to the directory inode. We can't delete the mountpoint directory as DM can because our data is there.

To make sure we're all on the same page, this overlay2 unmount function is called multiple times in the startup of a container.

It is also called for the pair of `docker stop; docker start`.

Docker expects data to persist between `stop/start` operations, so we can't just delete the upperdir on umount unless we ensure umount is only called when the container will really never be mounted/started again.

Clarification around what you're suggesting would be appreciated!

**Comment by cyphar** [comment]
@euank My suggestion is to replicate #34573 for overlayfs. This solves the problem as best as you can with modern kernels. As I've mentioned above, I am still working with kernel upstream to add new APIs that allow for killing mounts completely (even if they've been leaked to other namespaces) -- because currently users can attack Docker (or any other system that creates short-lived mounts) by creating an unprivileged mount namespace and then creating a bind-mount in a `tmpfs` that the host cannot see.

**Comment by thaJeztah** [comment]
ping @cpuguy83 @kolyshkin what's the status o this one?

**Comment by kolyshkin** [comment]
@thaJeztah I am looking into it, would love a test case here (which I might have tomorrow or so).

I found that removing mount point (i.e. `merged` directory) upon unmount also fixes the issue (but not prevent mounts to be leaked of course). I don't see why we can't implement both.

**Comment by euank** [comment]
Huh, deleting `merged` does dodge the warning/error. I have to take back my words above, I was under the impression that the inode tracking for workdir/upperdir (which we can't delete) would persist even if 'merged' changed, but that indeed is not the case. Thanks for the help and explanation @cyphar @kolyshkin, apologies for not understanding for a few there.

It should be safe enough to delete/recreate merged on each unmount/mount.

I'm not totally sure if it's the right thing to do though; effectively it's tricking the kernel into not emitting a warning which is still ultimately correct I think (that sharing a workdir isn't safe).

That being said, in practice we have problems with mount leaks, as far as I know we don't have actual problems with multiple active overlay mounts using the same dir, just multiple leaked-but-inactive ones.

I'll add on a commit to create/delete the merged dir.

> would love a test case here

Agreed, though it's non-trivial to make one, especially since I don't think the jenkins here has a new enough kernel to hit the failure case.

**Comment by euank** [comment]
Tossed on the mkdir/rmdir commit.

I've verified on my 4.13 kernel that either of the revert or the new 'remove merged' commit alone is enough to avoid the "workdir in use" warnings.

I agree with @kolyshkin that it's more correct to include both.

**Comment by amir73il** [comment]
@euank removing merged dir doesn't trick the kernel at all and it doesn't hide the warning either.
You were right to assume that overlayfs *will* honor exclusive workdir test across all mount namespaces,
because it does so at the dir inode level.

So how come removing "merged" dir fix the problem?
Because vfs_rmdir() calls  detach_mounts() before returning to userspace.
This call does lazy umount to all mounts in all mount namespaces which have the removed dir in their path.

In the case of "merged" dir, that means that all the leaked overlayfs mounts are lazily unmounted, which means that "sometime" after rmdir returns the leaked mount will be unmounted.
That "sometime" is actually deferred if the leaked mount is really being used by someone.

So removing "merged" is the real deal. It really provides docker genuine protection against a maliciously or irresponsibly leaked overlayfs mount that is being used.

The only problem is that you have no guaranty that the leaked mount will actually by unmounted by the time rmdir returns, although practically, I can't think of any reason for leaked unmount to be deferred if it is not being "used" (i.e. open file descriptor or current dir of a process inside it).

**Comment by euank** [comment]
Thanks a lot for clarifying! I'll update the commit message/comments with that understanding.

**Comment by cyphar** [comment]
@amir73il Your description is _mostly_ accurate (and this is why I did effectively the same thing in #34573). However,

> So removing "merged" is the real deal. It really provides docker genuine protection against a maliciously or irresponsibly leaked overlayfs mount that is being used.

There is another form of leak that there is currently no way of protecting against. Namely when the mount is pinned through bind-mounts (which are considered separate mounts of the underlying thing being bind-mounted). I've been working with kernel upstream to fix this particular issue (it looks like it'll require a new syscall).

**Comment by amir73il** [comment]
@cyphar you misread my statement about "real deal", or maybe I miswrote it...
What I meant is that removing "merged" is NOT a workaround.
It provides genuine protection from overlayfs POV, meaning workdir/upperdir are really never reused
by two different overlayfs mount without emitting a warning.
The reason that removing "merged" makes the warning go away in the bug is because this is the "good" sort of mount leak, which is already handled by kernel.
In the bind mount leak case you pointed out, overlayfs will still warn on workdir in use or refuse to mount depending on index=on/off configuration.

File: daemon/graphdriver/overlay/overlay.go
Comment (cyphar) [inline daemon/graphdriver/overlay/overlay.go]: Please undo this part of the patch, as discussed.

File: daemon/graphdriver/overlay2/overlay.go
Comment (cyphar) [inline daemon/graphdriver/overlay2/overlay.go]: And undo this part as well.

File: daemon/graphdriver/overlay/overlay.go
Comment (cyphar) [inline daemon/graphdriver/overlay/overlay.go]: This should be a `Remove` rather than a `RemoveAll`, and I think there should be a comment akin to the one in `devicemapper` explaining why we explicitly do a `Remove` here. The same applies to all the other places you made this change.

**Comment by kolyshkin** [comment]
I did a code review of last patch and left some comments but somehow made it so they are not displayed here. @euank please see this: https://github.com/moby/moby/commit/167b96d

**Comment by euank** [comment]
Intentionally leaving the `fixup!` unsigned to be sure I go through and autosquash it before this merges.

I think I've addressed or replied to the review comments.

File: daemon/graphdriver/overlay/overlay.go
Comment (kolyshkin) [reply daemon/graphdriver/overlay/overlay.go]: nit: `s/craeted/created/`

File: daemon/graphdriver/overlay2/overlay.go
Comment (kolyshkin) [inline daemon/graphdriver/overlay2/overlay.go]: ditto

**Review by kolyshkin — CHANGES_REQUESTED** [review]
LGTM except for a tiny nitpick

**Comment by kolyshkin** [comment]
Can you please rebase it?

**Comment by cpuguy83** [comment]
From the Z run:
> 22:52:17 error loading frozen images: ApplyLayer exit status 1 stdout:  stderr: unexpected EOF

Rebuilding.

**Comment by cyphar** [comment]
LGTM.

**Comment by thaJeztah** [comment]
ping @kolyshkin PTAL - nits were addressed

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by kolyshkin — APPROVED** [review]

**Comment by kolyshkin** [comment]
LGTM (disclaimer: I'm not a maintainer)

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks all for reviewing!

**Comment by doits** [comment]
Looks like this did not land in 17.11? I had to apply the patch manually, here's a [slightly modified version](https://gist.githubusercontent.com/doits/08ebc7122d0eededd1384b25f444a7bb/raw/dc9c9f0e0c704f3ac74c047877a5d5d8ab54cc2e/34948-17.11.diff) which applies correctly on 17.11 in case anyone else needs this.

**Comment by thaJeztah** [comment]
No it's not part of 17.11, will be in 17.12

**Pull request merged**
